accepted/2025/memory-safety/sdk-memory-safety-enforcement.md

```diff
@@ -0,0 +1,62 @@
+set of rules to use.
+
+## Memory Safety Rules Version
+
+The [metadata proposal](https://github.com/dotnet/csharplang/pull/9814) for C# details that we will stamp an attribute into modules that are compiled with
+the new rules enabled, `MemorySafetyRulesAttribute`. Enabling these new rules will be controlled at the csc level by a new flag: `/memorysafetyrules:number`,
+telling the compiler that version `number` of the memory safety rules should be used. This will then be controlled in the project file via the SDK, using a new
+property:
+
+```xml
+<MemorySafetyRules>_number_</MemorySafetyRules>
+```
+
+We propose using a number, rather than a simple on/off flag, as there may be future updates to the rules to tighten controls around ownership and use-after-free
+that aren't in the plan for .NET 11. We do not propose a `latest` or similar flag, as this an explicit breaking change that should be intentionally opted into,
+rather than letting it float. For .NET 11, we expect that this flag will be `preview`, as we do not expect that the feature will be ready for broad, unconditional
+adoption until .NET 12 or 13.
+
+### Open question: version number scheme
+
+Right now, there are two different approaches being considered for the version number. These approaches are:
+
+1. C# language version. This would follow the precedent set by `RefSafetyRulesAttribute`. Languages that want to interact with C#'s understanding of memory safety
+   rules will need to know and understand what that version number means to C#.
+2. .NET Runtime version. This would reflect the .NET-wide understanding of what memory safety means, and any languages (C# included) would need to know what the
+   runtime's safety guarantees are in order to interact with the setting.
+
+### AllowUnsafeBlocks
+
+The `AllowUnsafeBlocks` flag is related to this new feature, but is not directly impacted by it: the flag simply controls whether `unsafe` blocks are permitted
+in the program itself. We want projects to opt into the new memory safety rules, regardless of whether `AllowUnsafeBlocks` is set, so that they get appropriate
+errors when using the expanded definitions of `unsafe` APIs that properly covers things like `Unsafe.As`, the `Marshal` type, and others that hide memory
+unsafety behind `IntPtr`.
+
+## Default enablement
+
+In order to drive adoption of the feature, we will take the following approach:
+
+* In .NET 11, the feature will be off-by-default and in preview, and we will not be recommending broad adoption by arbitrary users until we have collected more
+  usability feedback and dialed in the enforcement. Users will be able to opt-in to the feature preview by putting `<MemorySafetyRules>preview</MemorySafetyRules>`
+  into their project files.
+* In the .NET 12 or 13 timeframe, once we have the ecosystem experience that we are fully confident in, we will turn the flag on-by-default for all projects that
```

Comment:
I still think that it's likely that we will want to opt off by default, but this is moot for .NET 11

Reply:
For sure. I wanted to document a PoV to avoid leaving our dear friend Andy to be the only one to share an option for the future.